lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php

```diff
@@ -0,0 +1,294 @@
+     *
+     * @throws InvalidArgumentException When the key name is invalid.
+     */
+    public function setDefaultKeyName($name)
+    {
+        $this->validateKeyName($name);
+        $this->defaultKeyName = $name;
+    }
+
+    /**
+     * Retrieves the default key name.
+     *
+     * @return string The default key name.
+     */
+    public function getDefaultKeyName()
+    {
+        return $this->defaultKeyName;
+    }
+
+    /**
+     * Sets a key name for a specific table.
+     *
+     * @param $table string The name of the table.
+     * @param $key string The name of the string.
+     *
+     * @throws InvalidArgumentException When the key or table name is invalid.
+     */
+    private function setKeyForTable($table, $key)
+    {
+        $this->validateTableName($table);
+        $this->validateKeyName($key);
+        $this->tableKeys[$table] = $key;
+    }
+
+    /**
+     * Retrieves a specific name for a key for a given table. The default is returned if this table does not have
+     * an actual override.
+     *
+     * @param string $tableName The name of the table.
+     *
+     * @return string
+     */
+    private function getKeyNameForTable($tableName)
+    {
+        return isset($this->tableKeys[$tableName]) ?
+            $this->tableKeys[$tableName] :
+            $this->defaultKeyName;
+    }
+
+    /**
+     * Prepares a key to be in a valid format for lookups for DynamoDB. If passing an array, that means that the key
+     * is the name of the key and the value is the actual value for the lookup.
+     *
+     * @param string $storageName
+     *
+     * @return array The key in DynamoDB format.
+     */
+    private function prepareKey($storageName, $key)
+    {
+        if (is_array($key)) {
+            $keyValue = reset($key);
+            $keyName = key($key);
+        } else {
+            $keyValue = $key;
+            $keyName = $this->getKeyNameForTable($storageName);
+        }
+
+        return $this->marshaler->marshalItem([$keyName => $keyValue]);
+    }
+
+    /**
+     * This is not true, but partial updates are too complicated given the available interface, meaning,
+     * the abstraction is insufficiently flexible enough to support this type of action.
+     *
+     * {@inheritdoc}
+     */
+    public function supportsPartialUpdates()
+    {
+        return false;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsCompositePrimaryKeys()
+    {
+        return true;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function requiresCompositePrimaryKeys()
+    {
+        return false;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function insert($storageName, $key, array $data)
+    {
+        $this->client->putItem([
+            self::TABLE_NAME_KEY => $storageName,
+            self::TABLE_ITEM_KEY => $this->prepareKey($storageName, $key) + $this->marshaler->marshalItem($data),
```

Comment:
DynamoDb chokes on empty values.
E.g 
```php
$data = [
  'foo' => '',
  'bar' => 'baz',
];
```
Would choke. The [existing implementation](https://github.com/doctrine/KeyValueStore/blob/master/lib/Doctrine/KeyValueStore/Storage/DynamoDbStorage.php#L194) includes a `prepareData` method - however it only consider one-dimensional arrays, multi-dimensional cleaning is needed too.

e.g
```php
$data = [
  'foo' => [
     'baz' => '',
     'bar' => 'boo',
  ],
  'bar' => 'baz',
];
```
needs to be sent as
```php
$data = [
  'foo' => [
     'bar' => 'boo',
  ],
  'bar' => 'baz',
];
```